ddtrace/tracer/handler.go

```diff
@@ -0,0 +1,235 @@
+			if err := h.prioritySampling.readRatesJSON(rc); err != nil {
+				h.config.statsd.Incr("datadog.tracer.decode_error", nil, 1)
+			}
+		}
+	}(h.payload)
+	h.payload = newPayload()
+}
+
+type hexInt uint64
+
+func (i hexInt) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + strconv.FormatUint(uint64(i), 16) + `"`), nil
+}
+
+type jsonSpan struct {
+	TraceID  hexInt             `json:"trace_id"`
+	SpanID   hexInt             `json:"span_id"`
+	ParentID hexInt             `json:"parent_id"`
+	Name     string             `json:"name"`
+	Resource string             `json:"resource"`
+	Error    int32              `json:"error"`
+	Meta     map[string]string  `json:"meta"`
+	Metrics  map[string]float64 `json:"metrics"`
+	Start    int64              `json:"start"`
+	Duration int64              `json:"duration"`
+	Service  string             `json:"service"`
+}
+
+type lambdaTraceWriter struct {
+	config    *config
+	payload   bytes.Buffer
+	hasTraces bool
+}
+
+var _ traceWriter = &lambdaTraceWriter{}
+
+func newLambdaTraceWriter(c *config) *lambdaTraceWriter {
+	w := &lambdaTraceWriter{
+		config: c,
+	}
+	w.resetPayload()
+	return w
+}
+
+const payloadLimit = 256 * 1024 // log line limit for cloudwatch
+
+func (h *lambdaTraceWriter) resetPayload() {
+	h.payload.Reset()
+	h.payload.WriteString(`{"traces": [`)
+	h.hasTraces = false
+}
+
+func (h *lambdaTraceWriter) addSpan(s *span) error {
+	js := jsonSpan{
+		TraceID:  hexInt(s.TraceID),
+		SpanID:   hexInt(s.SpanID),
+		ParentID: hexInt(s.ParentID),
+		Name:     s.Name,
+		Resource: s.Resource,
+		Error:    s.Error,
+		Meta:     s.Meta,
+		Metrics:  s.Metrics,
+		Start:    s.Start,
+		Duration: s.Duration,
+		Service:  s.Service,
+	}
+	e := json.NewEncoder(&h.payload)
+	err := e.Encode(js)
+	if err != nil {
+		return err
+	}
+	// cut trailing newline
+	h.payload.Truncate(h.payload.Len() - 1)
+	return nil
+}
+
+type encodingError struct {
+	cause      error
+	dropReason string
+}
+
+func (h *lambdaTraceWriter) addTrace(trace []*span) (int, *encodingError) {
+	startLen := h.payload.Len()
+	if !h.hasTraces {
+		h.payload.WriteByte('[')
+	} else {
+		h.payload.WriteString(", [")
+	}
+	written := 0
+	for i, s := range trace {
+		l := h.payload.Len()
+		if i != 0 {
+			h.payload.WriteByte(',')
+		}
+		err := h.addSpan(s)
+		if err != nil {
+			h.payload.Truncate(startLen)
+			return 0, &encodingError{cause: err, dropReason: "encoding_failed"}
+		}
+		if h.payload.Len() > payloadLimit-3 {
+			if i == 0 && !h.hasTraces {
+				// This is the first span of the first trace, and it's too big.
+				h.payload.Truncate(startLen)
+				return 0, &encodingError{cause: errors.New("span too large for payload"), dropReason: "trace_too_large"}
+			}
+			h.payload.Truncate(l)
+			break
+		}
+		written++
+	}
+	h.payload.WriteString("]")
+	h.hasTraces = true
+	return written, nil
+}
+
+func (h *lambdaTraceWriter) add(trace []*span) {
+	for len(trace) > 0 {
+		written, err := h.addTrace(trace)
```

Comment:
```suggestion
		n, err := h.writeTrace(trace)
```
Might be easier to understand due to the similarity to `io.Writer`